src/main/java/hudson/remoting/Engine.java

```diff
@@ -213,133 +296,147 @@ public void run() {
                 }
 
                 events.status("Locating server among " + candidateUrls);
-                Throwable firstError=null;
-                String host=null;
-                String port=null;
-                Set<String> agentProtocolNames = null;
-                SSLSocketFactory sslSocketFactory = getSSLSocketFactory();
-
-                for (URL url : candidateUrls) {
-                    String s = url.toExternalForm();
-                    if(!s.endsWith("/"))    s+='/';
-                    URL salURL = new URL(s+"tcpSlaveAgentListener/");
-
-                    // find out the TCP port
-                    HttpURLConnection con = (HttpURLConnection)Util.openURLConnection(salURL, credentials, proxyCredentials, sslSocketFactory);
-                    try {
-                        try {
-                            con.setConnectTimeout(30000);
-                            con.setReadTimeout(60000);
-                            con.connect();
-                        } catch (IOException x) {
-                            if (firstError == null) {
-                                firstError = new IOException("Failed to connect to " + salURL + ": " + x.getMessage()).initCause(x);
-                            }
+                final JnlpAgentEndpoint endpoint;
+                try {
+                    endpoint = resolver.resolve();
+                } catch (Exception e) {
+                    events.error(e);
+                    return;
+                }
+                if (endpoint == null) {
+                    events.status("Could not resolve server among " + candidateUrls);
+                    return;
+                }
+
+                events.status(String.format("Agent discovery successful%n"
```

Comment:
With JUL logging outputting all 4 lines as separate events is ugly and cramming the information into 1 line looks ugly too IMHO, so yes!